### DISCUSSION TITLE
Remove JSON support from System.BinaryData, or pull it entirely

### DISCUSSION THREAD

**Discussion by GSPP** [dotnet/runtime#68887] [General] [answered]
`System.BinaryData` (https://github.com/dotnet/runtime/blob/main/src/libraries/System.Memory.Data/src/System/BinaryData.cs), new to .NET 7, supports converting JSON to bytes. It was added here:

- https://github.com/dotnet/runtime/issues/41686

This seems like a pretty strong layering violation to me. The `System` namespace should not know the concept JSON.

JSON is not a future-proof format. XML seemed like the clear and eternal winner, and now is regarded a legacy format. JSON is not a really great format. It does its job alright. It is conceivable, though, that 10 years from now we will use something else.

The [design document](https://github.com/terrajobst/designs/blob/f114d834a3e67504a90b15e8551935427146a6c4/accepted/2020/binary-data/binary-data.md) acknowledges:

> Hence, most areas in the BCL wouldn't be able to take BinaryData, but technologies above the BCL could (Microsoft.Extensions.*, ASP.NET Core, Azure SDK etc). 

As designed, this type seems to live at a layer above the framework. It should be, perhaps, in the .NET web stack but not in the base class library.

Also, I have to say that I would not recommend using this type for anything. I'd recommend performing the appropriate conversion between data formats directly.

This new type is highly opinionated and specialized. I believe this type does not carry its weight and should be pulled entirely, or at least JSON support should be removed.

**Comment by GrabYourPitchforks** [comment, marked as answer]
The video (see https://github.com/dotnet/runtime/issues/41686#issuecomment-713056163) discusses this at length, since the "JSON isn't a future proof file format and isn't always the appropriate selection" concern is shared by many of the people who reviewed the API. The customer who wanted BinaryData also required JSON serialization support as a first class concept. The tight coupling is one of the reasons this type does not ship as part of the BCL and is a fully out of band package.

**Reply by GSPP** [reply]
Oh, it's out of band? I didn't notice that. Can you tell me how I can notice such things in the future? Like, where does it say what's shipped with the framework and what is shipped separately?

**Reply by jkoritzinsky** [reply]
When you go to the official documentation, you can see there. The "Applies to" section shows what is in box and what isn't. If the type is in ".NET Platform Extensions", then it's out-of-band. If it's in ".NET", then it's in the BCL.

In this repo, you can see which libraries are included in the [NetCoreAppLibrary.props file](https://github.com/dotnet/runtime/blob/main/src/libraries/NetCoreAppLibrary.props). If an assembly is listed in that file between the `NetCoreAppLibrary` tags, it is part of the BCL. If it is not, it is not in the BCL.
